flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/extraction/utils/DataTypeTemplate.java

```diff
@@ -0,0 +1,346 @@
+			return true;
+		}
+		if (o == null || getClass() != o.getClass()) {
+			return false;
+		}
+		DataTypeTemplate that = (DataTypeTemplate) o;
+		return Objects.equals(dataType, that.dataType) &&
+			Objects.equals(rawSerializer, that.rawSerializer) &&
+			Objects.equals(inputGroup, that.inputGroup) &&
+			version == that.version &&
+			Objects.equals(allowRawGlobally, that.allowRawGlobally) &&
+			Arrays.equals(allowRawPattern, that.allowRawPattern) &&
+			Arrays.equals(forceRawPattern, that.forceRawPattern) &&
+			Objects.equals(defaultDecimalPrecision, that.defaultDecimalPrecision) &&
+			Objects.equals(defaultDecimalScale, that.defaultDecimalScale) &&
+			Objects.equals(defaultYearPrecision, that.defaultYearPrecision) &&
+			Objects.equals(defaultSecondPrecision, that.defaultSecondPrecision);
+	}
+
+	@Override
+	public int hashCode() {
+		int result = Objects.hash(
+			dataType,
+			rawSerializer,
+			inputGroup,
+			version,
+			allowRawGlobally,
+			defaultDecimalPrecision,
+			defaultDecimalScale,
+			defaultYearPrecision,
+			defaultSecondPrecision);
+		result = 31 * result + Arrays.hashCode(allowRawPattern);
+		result = 31 * result + Arrays.hashCode(forceRawPattern);
+		return result;
+	}
+
+	// --------------------------------------------------------------------------------------------
+
+	@DataTypeHint
+	private static class DefaultAnnotationHelper {
+		// no implementation
+	}
+
+	private static DataTypeHint getDefaultAnnotation() {
+		return DefaultAnnotationHelper.class.getAnnotation(DataTypeHint.class);
+	}
+
+	private static <T> T defaultAsNull(DataTypeHint hint, Function<DataTypeHint, T> accessor) {
+		final T defaultValue = accessor.apply(DEFAULT_ANNOTATION);
+		final T actualValue = accessor.apply(hint);
+		if (Objects.deepEquals(defaultValue, actualValue)) {
+			return null;
+		}
+		return actualValue;
+	}
+
+	private static <T> T rightValueIfNotNull(T l, T r) {
+		if (r != null) {
+			return r;
+		}
+		return l;
+	}
+
+	private static Boolean hintFlagToBoolean(HintFlag flag) {
+		if (flag == null) {
+			return null;
+		}
+		return flag == HintFlag.TRUE;
+	}
+
+	private static DataType extractDataType(
+			DataTypeLookup lookup,
+			@Nullable String typeName,
+			@Nullable Class<?> conversionClass,
+			DataTypeTemplate template) {
+		// explicit data type
+		if (typeName != null) {
+			// RAW type
+			if (typeName.equals(RAW_TYPE_NAME)) {
+				return createRawType(lookup, template.rawSerializer, conversionClass);
+			}
+			// regular type that must be resolvable
+			return lookup.lookupDataType(typeName)
+				.map(dataType -> {
+					if (conversionClass != null) {
+						return dataType.bridgedTo(conversionClass);
+					}
+					return dataType;
+				})
+				.orElseGet(() -> {
+					throw extractionError("Could not resolve type with name '%s'.", typeName);
+				});
+		}
+		// extracted data type
+		else if (conversionClass != null) {
+			return DataTypeExtractor.extractFromType(lookup, template, conversionClass);
```

Comment:
This case is not covered by tests. I think it's worth adding one, as it is quite different from the rest.

```
public @DataTypeHint(bridgedTo = long.class) Object primitiveLongField; //-> should be extracted as BIGINT NOT NULL
```